### PR TITLE
Handle dict of metrics in distributed training

### PR DESCRIPTION
**Pull request by froody**
When training with a model with a subclass `tf.keras.metrics.Metric` which which returns a dict, e.g.

```
class CustomMetric(tf.keras.metrics.Metric):
    def result(self):
        return {"good", tf.convert_to_tensor([1.0]), "bad": tf.convert_to_tensor([0.0])}
```

I got an exception in `merge_fn_wrapper` when trying to call `array_ops.identity` on a `Dict[str, tf.Tensor]`. This fix allows me to do distributed training with the above metric.

### CODE REVIEW

File: keras/utils/metrics_utils.py [hunk 1]
```diff
@@ -174,7 +174,13 @@ def merge_fn_wrapper(distribution, merge_fn, *args):
                 # Wrapping result in identity so that control dependency between
                 # update_op from `update_state` and result works in case result
                 # returns a tensor.
-                return tf.identity(result)
+                if isinstance(result, dict):
```
Comment (qlzh727) [inline keras/utils/metrics_utils.py:177]: I think the logic between L140 and L157 can be reuse here, how about extract that into a function and reuse it?

File: keras/utils/metrics_utils.py [hunk 2]
```diff
@@ -174,7 +174,13 @@ def merge_fn_wrapper(distribution, merge_fn, *args):
...
+                    return {
+                        key: array_ops.identity(value)
```
Comment (qlzh727) [inline keras/utils/metrics_utils.py:179]: Note that keras use the public TF API. This should be tf.identity(), and array_ops is a private TF API.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
Hi @froody Can you please check @qlzh727's comments and resolve conflicts?. Thank you!

**Comment by gbaned** [comment]
Hi @froody Any update on this PR? Please. Thank you!
